R/time_bins.R

```diff
@@ -0,0 +1,206 @@
+#' @section Developer(s):
+#' Lewis A. Jones
+#' @section Reviewer(s):
+#' Kilian Eichenseer
+#' @examples
+#' #Using interval midpoint age
+#' time_bins(interval = 10, plot = TRUE)
+#'
+#' #Using interval age range
+#' time_bins(interval = c(50, 100), plot = TRUE)
+#'
+#' #Using a single interval name
+#' time_bins(interval = c("Maastrichtian"), plot = TRUE)
+#'
+#' #Using a range of intervals defined by two named intervals and equal duration bins
+#' time_bins(interval = c("Fortunian", "Meghalayan"), size = 10, plot = TRUE)
+#'
+#' #Assign bins based on given age estimates
+#' time_bins(interval = c("Fortunian", "Meghalayan"), assign = c(232, 167, 33), plot = TRUE)
+#' @export
+time_bins <- function(interval = c("Fortunian", "Meghalayan"), size = NULL, assign = NULL, scale = "GTS2020", plot = FALSE){
+  #error handling
+  if (!is.character(interval) & !is.numeric(interval)) {
+    stop("`interval` must be of class 'character' or 'numeric'")
+  }
+
+  if (is.numeric(size) == FALSE & is.null(size) == FALSE) {
+    stop("Size should be a numeric or NULL")
+  }
+
+  if (is.logical(plot) == FALSE) {
+    stop("plot should be logical (TRUE/FALSE)")
+  }
+
+  if (is.numeric(assign) == TRUE & any(assign < 0) == TRUE) {
+    stop("Age estimates should be non-negative values.
+  You can transform your data using abs().")
+  }
+
+  if (scale != "GTS2012" & scale != "GTS2020") {
+    stop("Scale must be either 'GTS2012' or 'GTS2020'")
+  }
+
+  #grab data
+
+  #which geological timescale to use?
+
+  if(scale == "GTS2020"){df <- palaeoverse::GTS2020}
+  if(scale == "GTS2012"){df <- palaeoverse::GTS2012}
+
+  #which rank is required? Non-stage level is only available for character string.
+  if(is.character(interval)){
+    rnk <- df[which(df$interval_name %in% interval), c("rank")]
+
+    if(length(rnk) != length(interval)){stop(paste0("Checking spelling of specified intervals.
+    Available intervals are accessible via GTS2020$interval_name and GTS2012$interval_name."))}
+
+    rnk <- unique(rnk)}else{rnk <- "Stage"}
+
+  if(length(rnk) != 1){stop("Interval ranks should be the same, e.g. specifiy only stage or period names, not a combination.")}
+
+  #subset to rank
+  df <- subset(df, rank == rnk)
+
+  #character string entered
+  if(is.character(interval)){
+    if(length(interval) == 1){
+      w <- which(df$interval_name %in% interval)
+      if(length(w) != length(interval)){stop(paste0("Checking spelling of specified intervals.
+    Available intervals are accessible via GTS2020$interval_name."))}
+      df <- df[w,]
+    }
+    if(length(interval) == 2){
+      w <- which(df$interval_name %in% interval)
+      df <- df[w[1]:w[2],]
+    }
+  }
+  #numeric ages entered
+  if(is.numeric(interval)){
+    if(length(interval) == 1){
+      if(interval > max(df$max_ma) | interval < min(df$min_ma)){
+        stop("Value does not appear in range of available intervals: 0 to 541")
+      }
+      df <- df[which.min(abs(df$mid_ma - interval)),]
```

Comment:
df <- df[which.min(abs(df$mid_ma - interval)),]

should be changed to 
      int_index <- which(interval < df$max_ma & interval >= df$min_ma)
      df <- df[int_index,]

as current version is not giving the right output in some cases, e.g. time_bins(20) returns the Aquitanian (20.44 - 23.03 Ma).

Reply:
Thanks Kilian, this has now been implemented. Originally I was going with the closest midpoint, but this makes way more sense.

Regarding the behaviour when the numeric input is on the boundary, I have updated this to return both intervals. I think this is the most transparent for the user. They can then make a decision if they want to update their input age.